keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1150 @@
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+All Kubernetes APIs support [field
+selection](https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/)
+for metadata.name, metadata.namespace.
+
+Additionally, built-in APIs support field selection on specific fields such as
+status.phase for Pods, but CustomResourceDefinitions lack equivalent
+functionality.
+
+Without this enhancement, an extension author that wants to be be able to filter
+resources might choose to:
+
+  - Organize data into a label even though it should be a spec or status field.
+  - Double writing data both to a label and a spec or status field.
+
+Both of these should be avoided. We should enable extension authors structure
+their APIs according to best practices, not based on data access patterns.
+
+### Goals
+
+- Add an allow list of selectable fields to CustomResourceDefinitions.
+- Performance and resource consumption characteristics are roughtly equivalent
+  to label selectors.
+
+### Non-Goals
+
+- Arbitrary field selection.
+   - This proposal is to support field selection for a small allow list of
+     fields. Expanding support to include all fields is complicated by:
+     - Fields contained in, or nested under, lists or maps.
+     - The costs to apiserver of evaluating filters on objects or maintaining
+       indices.
+     - Lack of direct support for filtering in etcd.
+- More sophisticated field selection logic. Possible approaches include: CEL
+  based field selection, enhancements to field selection such as set based
+  operators like provided for labels ("in", "not in"), OR clauses, pattern
+  matching... See the "CEL as an accelerated predicate language" section "Future
+  Work" for more details on what makes this challenging.
+
+
+## Proposal
+
+Add `selectableFields` to the versions of CustomResourceDefinition:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: selector.stable.example.com
+spec:
+  group: stable.example.com
+  versions:
+    - name: v1
+      selectableFields:
+      - jsonPath: .spec.color 
```

Comment:
@jpbetz today we allow expressions to be evaluated on an instance before evaluation. For instance, `Itoa` for replicasets https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/apps/replicaset/strategy.go#L184-L190.  Conceptually this could do anything, but I don't recall anyone taking advantage of it in kube.

Will we support casting, but nothing else?  Do you ever see this expanding?  An explicit non-goal could put it to rest.

Reply:
Yes, https://github.com/kubernetes/enhancements/blob/1504d0aaba219a18a0b3c2f9c17940cf0a7e82a0/keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md#background-field-selection-as-it-exists-before-this-enhancement summarizes all the casts performed for built-in types. Note that all casts are to string.

I'd like to keep it this simple since all --field-selector matching uses string equality checking.  I'll make sure this is clear in the non-goals.

Also, I've limited which field types can be used in CRDs to strings, enums, integers and booleans to match what we support for built-ins (we also support enums and UID, but in CRDs those are represented as strings with formats and would be allowed).